scripts/gas-refund-program/transactions-indexing/transaction-resolver.ts

```diff
@@ -0,0 +1,156 @@
+ */
+import { STAKING_POOL_ADDRESSES } from '../../../src/lib/gas-refund'
+import { covalentGetTXsForContract } from './txs-covalent'
+import { getTransactionGasUsed } from '../staking/covalent';
+import { getSuccessfulSwaps } from './swaps-subgraph'
+import { GasRefundTransaction, CovalentTransaction } from '../types'
+import { GasRefundTxOriginCheckStartEpoch, GasRefundSwapSourceCovalentStartEpoch, AUGUSTUS_ADDRESS } from '../../../src/lib/gas-refund'
+
+type GetAllTXsInput = {
+  startTimestamp: number;
+  endTimestamp: number;
+  chainId: number;
+  epoch: number;
+}
+
+export const getAllTXs = async ({epoch, chainId, startTimestamp, endTimestamp}: GetAllTXsInput): Promise<GasRefundTransaction[]> => {
+
+  // fetch swaps and stakes
+  const allTXs = await Promise.all([
+    getSwapTXs({epoch, chainId, startTimestamp, endTimestamp}),
+    getStakingTXs({chainId, startTimestamp, endTimestamp})
+  ])
+
+  const allTXsFlattened = [].concat.apply([], allTXs) as GasRefundTransaction[]
+
+  // sort to be chronological
+  const allTXsChronological = allTXsFlattened.sort((a, b) => +(a.timestamp) - +(b.timestamp));
+
+  return allTXsChronological
+}
+
+
+/**
+ * this will take an epoch, a chain, and two timespan values (min/max).
+ * it will use subgraph for now (and augment gas data via a covalent call),
+ * but later resolve to covalent after a certain epoch.
+*/
+type GetSwapTXsInput = {
+  startTimestamp: number;
+  endTimestamp: number;
+  chainId: number;
+  epoch: number;
+}
+export const getSwapTXs = async ({ epoch, chainId, startTimestamp, endTimestamp }: GetSwapTXsInput): Promise<GasRefundTransaction[]> => {
+  const swaps: GasRefundTransaction[] = await (async () => {
+    // todo: epoch check when we change over - remove `false &&`
+    if (false && epoch >= GasRefundSwapSourceCovalentStartEpoch) {
+      // get from covalent
+      const swapsFromCovalent = await covalentGetTXsForContract({
+        startTimestamp,
+        endTimestamp,
+        chainId,
+        contract: AUGUSTUS_ADDRESS
+      })
+      const normalisedSwapsFromCovalent = swapsFromCovalent.map(swap => ({
+        ...swap,
+        blockNumber: swap.blockNumber.toString()
+      }))
+      return normalisedSwapsFromCovalent
+    } else {
+      // get swaps from the graph
+      const swaps = await getSuccessfulSwaps({ startTimestamp, endTimestamp, chainId })
+
+      // optionally filter out smart contract wallets
+      const filteredSwaps = swaps.filter(swap =>
+        epoch < GasRefundTxOriginCheckStartEpoch ||
+        epoch >= GasRefundTxOriginCheckStartEpoch &&
+        swap.initiator !== swap.txOrigin
+      )
+
+      // augment with gas used
+      const swapsWithGasUsedNormalised: GasRefundTransaction[] = await Promise.all(
+        filteredSwaps.map(async ({
+          txHash,
+          txOrigin,
+          txGasPrice,
+          timestamp,
+          blockNumber
+        }) => {
+          const txGasUsed = await getTransactionGasUsed({
```

Comment:
The problem I see here is that we are going to call covalent to every tx no matter whether it’s eligible for refund or not which was more frugal in previous version.

Reply:
ah yes, good point